File: remotestate-ts/src/lib/protocol.ts
import type { Path } from "./path";

/**
 * One changed store value.
 */
export interface StateUpdate {
  /**
   * Normalized path into the state.
   */
  path: Path;

  /**
   * Value at the path.
   */
  value: unknown;
}

// -------------------------------------------
// JS --> Python
// -------------------------------------------

/**
 * Request the current value at one store path.
 */
export interface GetMessage {
  /**
   * Protocol discriminator for store get requests.
   */
  type: "get";

  /**
   * Internal call ID used to correlate the response.
   */
  call_id: string;

  /**
   * State path to fetch.
   */
  path: Path;
}

/**
 * Write one value at a store path.
 */
export interface SetMessage {
  /**
   * Protocol discriminator for store set requests.
   */
  type: "set";

  /**
   * Internal call ID used to correlate the response.
   */
  call_id: string;

  /**
   * State path to write.
   */
  path: Path;

  /**
   * Value to assign at the path.
   */
  value: unknown;
}

/**
 * Invoke a state-mutating service method.
 */
export interface ActionMessage {
  /**
   * Protocol discriminator for action calls.
   */
  type: "action";

  /**
   * Internal call ID used to correlate the response.
   */
  call_id: string;

  /**
   * Optional user-supplied task ID for progress tracking.
   */
  task_id?: string;

  /**
   * Service action method name.
   */
  method: string;

  /**
   * Positional arguments passed to the action.
   */
  args: unknown[];

  /**
   * Keyword arguments passed to the action.
   */
  kwargs: Record<string, unknown>;
}

/**
 * Invoke a read-only service method that returns a value.
 */
export interface QueryMessage {
  /**
   * Protocol discriminator for query calls.
   */
  type: "query";

  /**
   * Internal call ID used to correlate the response.
   */
  call_id: string;

  /**
   * Optional user-supplied task ID for progress tracking.
   */
  task_id?: string;

  /**
   * Service query method name.
   */
  method: string;

  /**
   * Positional arguments passed to the query.
   */
  args: unknown[];

  /**
   * Keyword arguments passed to the query.
   */
  kwargs: Record<string, unknown>;
}

// -------------------------------------------
// Python --> JS
// -------------------------------------------

/**
 * Return one value requested by a previous `GetMessage`.
 */
export interface GetResultMessage {
  /**
   * Protocol discriminator for store get results.
   */
  type: "get_result";

  /**
   * Internal call ID from the request.
   */
  call_id: string;

  /**
   * State path that was fetched.
   */
  path: Path;

  /**
   * Value at the requested path.
   */
  value: unknown;
}

/**
 * Return the batched store updates produced by an action.
 */
export interface ActionResultMessage {
  /**
   * Protocol discriminator for action results.
   */
  type: "action_result";

  /**
   * Internal call ID from the request.
   */
  call_id: string;

  /**
   * Changed state values.
   */
  updates: StateUpdate[];
}

/**
 * Return the batched store updates produced by a store set request.
 */
export interface SetResultMessage {
  /**
   * Protocol discriminator for store set results.
   */
  type: "set_result";

  /**
   * Internal call ID from the request.
   */
  call_id: string;

  /**
   * Changed state values.
   */
  updates: StateUpdate[];
}

/**
 * Return the result of a previous `QueryMessage`.
 */
export interface QueryResultMessage {
  /**
   * Protocol discriminator for query results.
   */
  type: "query_result";

  /**
   * Internal call ID from the request.
   */
  call_id: string;

  /**
   * Value returned by the query.
   */
  value: unknown;
}

/**
 * Report progress for a tracked action or query.
 */
export interface TaskUpdateMessage {
  /**
   * Protocol discriminator for task progress updates.
   */
  type: "update_task";

  /**
   * Internal call ID for the running task.
   */
  call_id: string;

  /**
   * User-supplied task ID.
   */
  task_id: string;

  /**
   * Service method associated with the task.
   */
  method: string;

  /**
   * Current task status.
   */
  status: "running" | "done" | "error";

  /**
   * Optional short task name for display.
   */
  name?: string;

  /**
   * Optional task detail for display.
   */
  detail?: string;

  /**
   * Optional progress percentage from 0 to 100.
   */
  progress?: number; // 0-100

  /**
   * Optional error message when the task is in the `error` state.
   */
  error?: string;
}

/**
 * Return an error for a previous request.
 */
export interface ErrorMessage {
  /**
   * Protocol discriminator for errors.
   */
  type: "error";

  /**
   * Internal call ID from the failed request.
   */
  call_id: string;

  /**
   * Error message returned by Python.
   */
  message: string;
}

// -------------------------------------------
// Derived
// -------------------------------------------

/**
 * Any message the Remote State bridge can send to Python.
 */
export type IncomingMessage =
  | GetMessage
  | SetMessage
  | ActionMessage
  | QueryMessage;

/**
 * Any message Python can send back to the Remote State bridge.
 */
export type OutgoingMessage =
  | GetResultMessage
  | SetResultMessage
  | ActionResultMessage
  | QueryResultMessage
  | TaskUpdateMessage
  | ErrorMessage;

File: remotestate-py/src/remotestate/protocol.py
from collections.abc import Sequence
from typing import Annotated, Any, Literal

from pydantic import BaseModel, BeforeValidator, Field

from .path import Path, PathSegmentInput, normalize_path


def _normalize_protocol_path(path: Sequence[PathSegmentInput]) -> Path:
    if isinstance(path, str):
        raise ValueError("RemoteState protocol paths must be arrays of segments")
    try:
        return normalize_path(path)
    except TypeError as e:
        raise ValueError(str(e)) from e


ProtocolPath = Annotated[Path, BeforeValidator(_normalize_protocol_path)]


class StateUpdate(BaseModel):
    """One changed store value."""

    path: ProtocolPath
    """Normalized path into the store's state."""

    value: Any
    """The JSON value at ``path``."""


# ----------------------------------------------------
# JS --> Python
# ----------------------------------------------------


class GetMessage(BaseModel):
    """Request one store value by path."""

    type: Literal["get"] = "get"
    """Message type."""

    call_id: str
    """An internal get-ID."""

    path: ProtocolPath
    """Normalized path into the store's state."""


class SetMessage(BaseModel):
    """Set one store value by path."""

    type: Literal["set"] = "set"
    """Message type."""

    call_id: str
    """An internal set-ID."""

    path: ProtocolPath
    """Normalized path into the store's state."""

    value: Any
    """New value to assign at ``path``."""


class ActionMessage(BaseModel):
    """Invoke a service action.

    A matching ``ActionResultMessage`` carries the resulting state changes.
    """

    type: Literal["action"] = "action"
    """Message type."""

    call_id: str
    """An internal action-ID."""

    task_id: str | None = None
    """User-supplied task identifier for status and progress tracking."""

    method: str
    """The action method's name."""

    args: list[Any] = []
    """The action method's positional arguments."""

    kwargs: dict[str, Any] = {}
    """The action method's keyword arguments."""


class QueryMessage(BaseModel):
    """Invoke a read-only service query.

    Queries return a value via ``QueryResultMessage`` and may optionally carry
    a task ID when the caller wants progress updates.
    """

    type: Literal["query"] = "query"
    """Message type."""

    call_id: str
    """An internal query-ID."""

    task_id: str | None = None
    """User-supplied task identifier for status and progress tracking."""

    method: str
    """The query method's name."""

    args: list[Any] = []
    """The query method's positional arguments."""

    kwargs: dict[str, Any] = {}
    """The query method's keyword arguments."""


# ----------------------------------------------------
# Python --> JS
# ----------------------------------------------------


class GetResultMessage(BaseModel):
    """Return the current value for a previous ``GetMessage``."""

    type: Literal["get_result"] = "get_result"
    """Message type."""

    call_id: str
    """An internal get-ID."""

    path: ProtocolPath
    """Normalized path into the store's state."""

    value: Any
    """The JSON value of the state value."""


class ActionResultMessage(BaseModel):
    """Return the batched store updates produced by a previous ``ActionMessage``."""

    type: Literal["action_result"] = "action_result"
    """Message type."""

    call_id: str
    """An internal action-ID."""

    updates: list[StateUpdate]
    """Changed state values. May be empty."""


class SetResultMessage(BaseModel):
    """Return the batched store updates produced by a previous ``SetMessage``."""

    type: Literal["set_result"] = "set_result"
    """Message type."""

    call_id: str
    """An internal set-ID."""

    updates: list[StateUpdate]
    """Changed state values."""


class QueryResultMessage(BaseModel):
    """Return the computed result for a previous ``QueryMessage``."""

    type: Literal["query_result"] = "query_result"
    """Message type."""

    call_id: str
    """An internal query-ID."""

    value: Any
    """The query result."""


class TaskUpdateMessage(BaseModel):
    """Report task progress for a tracked action or query.

    These messages are only emitted when the caller supplied a task ID.
    """

    type: Literal["update_task"] = "update_task"
    """Message type."""

    call_id: str
    """An internal action- or query-ID."""

    task_id: str
    """User-supplied task identifier."""

    method: str
    """The method name."""

    status: Literal["running", "done", "error"]
    """Task status."""

    name: str | None = None
    """Task name."""

    detail: str | None = None
    """Task detail text."""

    progress: float | None = Field(None, ge=0, le=100)
    """Task progress, a number between 0 and 100."""

    error: str | None = None
    """Error message. Valid only if status is `"error"`."""


class ErrorMessage(BaseModel):
    """Return an error for a previous request or parse failure."""

    type: Literal["error"] = "error"
    """Message type."""

    call_id: str
    """An internal request ID."""

    message: str
    """Error message text."""


# ----------------------------------------------------
# Discriminated Unions for the Dispatcher
# ----------------------------------------------------

IncomingMessage = Annotated[
    GetMessage | SetMessage | ActionMessage | QueryMessage,
    Field(discriminator="type"),
]
"""Any message that can be sent from JavaScript to Python."""

OutgoingMessage = Annotated[
    GetResultMessage
    | SetResultMessage
    | QueryResultMessage
    | TaskUpdateMessage
    | ActionResultMessage
    | ErrorMessage,
    Field(discriminator="type"),
]
"""Any message that can be sent from Python back to JavaScript."""
